Feat/3533 arbitrary contract interaction

Nortsova opened the pull request:
## Description
This issue is to add the intended functionality to the modal. Upon user entering the address, it should fetch the contract ABI ([from Arbiscan API](https://docs.arbiscan.io/api-endpoints/contracts#get-contract-abi-for-verified-contract-source-codes)) and:

Display the ABI in an editable text field
Allow the user to select which function from the ABI they want to interact with
For the selected function, display fields for each of the function's arguments that the user can edit

Figma: https://www.figma.com/design/5V8pr7iMwXsT9L3VAZsmUt/Actions-%26-Motions?node-id=24968-204505&node-type=frame&t=VK5oE9wW19JFFa0k-0 

- [x] Both desktop and mobile views adhere to Figma designs
- [x]  Target contract address is validated for its correctness before sending network calls (all API plans have calls/second restrictions so this is important)
- [x]  Fetched ABI is displayed in the ABI/JSON field, or
- [x]  Error is displayed on ABI fetching failure
- [x]  A confirmation overlay is shown before allowing the user to manually edit the ABI ([Figma](https://www.figma.com/design/5V8pr7iMwXsT9L3VAZsmUt/Actions-%26-Motions?node-id=26568-211345&t=VK5oE9wW19JFFa0k-4))
- [x]  User-edited ABI is validated (checking JSON validity at least; need to confirm if there's a feasible way of validating ABI formatting)
- [x]  On successful ABI parsing, dropdown is shown displaying available contract functions
- [x]  On selecting a function, list of fields matching its arguments is shown. These should be editable by the user
- [x]  Function argument fields are validated according to the rules below

## Testing

> [!NOTE]  
>  Add ARBISCAN_API=https://api.arbiscan.io/api to .env.secrets. For ARBISCAN_API_KEY, contact me or create a personal key [here](https://arbiscan.io/myapikey).


Step 1. Open "Arbitrary transaction" Motion (Open any motion and then select "Arbitrary transaction" from dropdown) 
Step 2. Click "Add transaction" and verify the modal matches the design.
Step 3. Verify that fields are required:
<img width="537" alt="image" src="https://github.com/user-attachments/assets/99222143-b4a9-4ebd-a87d-e6f4cdb6a8bd">

Step 4. Enter an invalid contract address; verify the "Please enter a valid contract address." error appears
<img width="504" alt="image" src="https://github.com/user-attachments/assets/aa9ab6fe-229d-46ed-864f-4992a329fb8a">
 
Step 5. Enter an unverified contract address: `0xb77D57F4959eAfA0339424b83FcFaf9c15407461` and verify the server error:
<img width="499" alt="image" src="https://github.com/user-attachments/assets/1274cd7e-eadd-4680-97c4-17fa1a914fdc">

Step 6. Enter an empty contract address (`0x272632C8C7e34957CB13F9E2De5F2EA5628F818A`) and verify the error "Contract ABI doesn't have any available methods." :
<img width="636" alt="image" src="https://github.com/user-attachments/assets/6a1c8cdc-c9e2-4e52-9e7f-6531e319761b">

Step 7. Hover over the JSON ABI and verify overlay visibility; click "Edit generated ABI" to hide the overlay.
<img width="382" alt="image" src="https://github.com/user-attachments/assets/75e7877f-6ead-478e-a7d4-0c420e89d49d">

Step 8. Enter a correct contract address (`0xeeeafa86903cf2cb5c74a61d3e67844e226730c0`); verify the JSON ABI loads correctly:
<img width="402" alt="image" src="https://github.com/user-attachments/assets/3d92865a-3435-4dcd-bc5b-f8fc22a93fec">

Step 9. Verify the overlay visibility for this contract as well:
<img width="394" alt="image" src="https://github.com/user-attachments/assets/746a2e7e-3bbd-489b-b153-cf550fc3e0aa">

Step 10. Submit the form and confirm validation works, as shown below:
<img width="395" alt="image" src="https://github.com/user-attachments/assets/578a956a-31a9-461f-a92a-1f0a9464f8b4">

Step 11. Select the "approve" method, verify that the **to (address)** and **tokenId (unit256)** field appears, and validate it:
<img width="393" alt="image" src="https://github.com/user-attachments/assets/44deb6d9-3bab-451d-a09c-1ee0ebe9314d">

Step 12. Ensure the **to address** field accepts only valid addresses:
<img width="369" alt="image" src="https://github.com/user-attachments/assets/5e51e8ae-43bf-4a7a-90d2-ad0d14a00e26">


Step 13. Modify the ABI JSON, updating the "approve" method types to check validation functionality per this task’s requirements.

### Function arguments validation

| Argument Type						| Validation rules									| 
| --                        						| --												|
| int{m} where m is between 8 and 256 	| Integer between -2^(m-1) and 2^(m-1) - 1				| 
| uint{m} where m is between 8 and 256	| Integer between Integer between 0 and 2^m - 1		|
| address								| Standard address rules, could use `isAddress` from ethers |
| bool | One of: "true", "false", 0 or 1 | 
| bytes32 | 64 hex chars (0-9, A-F), must start with `0x` |
| bytes | Even number of hex chars, must start with `0x` |

Additionally, arguments could be one or two dimensional arrays `{type}[] and {type}[][], for example int256[]`.


Step 14. Submit the form and confirm values display in the table as shown:
<img width="559" alt="image" src="https://github.com/user-attachments/assets/2fe9288b-68fe-4e3d-b88e-4c201e4107b8">

### Updates after review:
- `fetchArbiscanContract` deleted and replaced with `fetchContractABI`
- Error formatting and server error handling fixed:
<img width="476" alt="image" src="https://github.com/user-attachments/assets/b3470482-2455-4be4-9d43-e700fbe75b05">

- Confirmation modal added:
<img width="561" alt="image" src="https://github.com/user-attachments/assets/35828d15-09fc-465b-9cf1-27649e98c204">

- Modal buttons fixed for mobile view (Previously, buttons were not sticking to the bottom)
<img width="617" alt="image" src="https://github.com/user-attachments/assets/f1935d65-d739-4575-99a1-043ca315486b">

- Default state for the table set to empty table (This was based on Kerry’s comment https://github.com/JoinColony/colonyCDapp/pull/3624#pullrequestreview-2435351878 and a discussion with @melyndav )
<img width="699" alt="image" src="https://github.com/user-attachments/assets/5ed32e07-7d6a-4fe2-9179-e08abf242f7f">

-Dynamic validation bug fixed (as noted by @iamsamgibbs ):
![Nov-18-2024 21-51-52](https://github.com/user-attachments/assets/9a64582c-f6ea-4056-8571-30b4559c1a18)

- Form fields now clear correctly when the method is changed:
<img width="700" alt="image" src="https://github.com/user-attachments/assets/d369f037-7c76-49c9-8510-c7230750dd4e">

-Spinner added for contract loading:
<img width="465" alt="image" src="https://github.com/user-attachments/assets/c30f23ea-2cf3-4d91-b9f5-9447ca3f1c51">

- Format JSON / Unformat JSON text was added: 
<img width="468" alt="image" src="https://github.com/user-attachments/assets/14fc153d-bd6c-4ea3-9840-386285bdbda1">

- Improved filtering. Now only functions (not constants) are displayed:
<img width="517" alt="image" src="https://github.com/user-attachments/assets/dc754a1b-fcc2-44e7-9199-f74589d1f750">



## Diffs

**Changes** 🏗

* I updated Textarea, FormInput and FormSelect to accept inline validation.
* Textarea now has correct error message and overlay


Resolves #3533

iamsamgibbs commented on src/components/v5/common/ActionSidebar/partials/forms/ArbitraryTxsForm/partials/AddTransactionModal/ContractAddressInput.tsx:
```diff
@@ -0,0 +1,48 @@
+  const [serverError, setServerError] = useState('');
+
+  useEffect(() => {
+    if (isAddress(contractAddressField)) {
+      getABIFromContractAddress(contractAddressField);
+    }
+
```
Could do:

```
if (!isAddress(contractAddressField)) {
return;
}

const response = await fetchArbiscanContract(contractAddress)
...
```

iamsamgibbs commented on src/components/v5/common/ActionSidebar/partials/forms/ArbitraryTxsForm/partials/AddTransactionModal/ContractAddressInput.tsx:
```diff
@@ -0,0 +1,48 @@
+    async function getABIFromContractAddress(contractAddress: string) {
+      const response = await fetchArbiscanContract(contractAddress);
+
+      if (response.status === '0') {
+        setServerError(response.result);
```
This didn't catch if I didn't have the API key set.

Nortsova replied:
Hey, seems that this one is expected from API. It allow you to send couple of requests without API key:

<img width="489" alt="image" src="https://github.com/user-attachments/assets/587f22d8-6a00-4537-a3a8-0390b1d06e6e">

iamsamgibbs replied:
Oh, in that case it wasn't showing the error for invalid address in that scenario (I was using the colony address)

Nortsova replied:
Yes, you’re right. Even though the API allows us to make calls without an API key (with certain limits), there was definitely a bug with the server validation. I’ve fixed it, and I hope the issue you mentioned will now be resolved with these changes. 🙌

iamsamgibbs reviewed (changes requested):
Great work on this complicated form so far! Dynamic forms are always a bit of a pain!

This is mostly working great, but I have hit a few small issues.

I think this modal needs some kind of cancel confirmation if you click outside of the modal. Currently it just closes so if you are halfway through editing and accidentally click outside of the modal, you lose everything you've done so far.

The validation error here should have proper formatting.
 
<img width="467" alt="Screenshot 2024-11-14 at 11 49 18" src="https://github.com/user-attachments/assets/97061d13-0e44-4e61-99e2-d66c8f81b2d5">

The validation on the dynamic fields doesn't update until the value in the field is changed. Here I've updated the ABI to accept address but it is still validating for bool.

<img width="472" alt="Screenshot 2024-11-14 at 11 58 27" src="https://github.com/user-attachments/assets/3ab8e382-94a1-4154-91b4-537cf3cd1b4b">

It seems like the fields are preserved when switching between the different methods. This means if you switch between different methods, then submit, you will end up with extra empty fields in the table.
 
<img width="667" alt="Screenshot 2024-11-14 at 12 00 55" src="https://github.com/user-attachments/assets/f85525a9-a095-4181-9682-2b5b2d536de2">

It might be nice to show some kind of loading state whilst it is fetching arbiscan, on slower connections it doesn't look like anything is happening.

https://github.com/user-attachments/assets/a984283e-4eeb-4552-8adb-9355dad239a9

I also didn't get any kind of error message when trying this without an API key, so I'm not sure if it is handling error responses properly.

Overall though this is great work so far, a lot of the functionality is there and working great already.

jakubcolony reviewed (changes requested):
Great start on this, the bulk of the functionality is working really nice, and I only picked up on some minor issues 👍 

Error messages should be capitalized, maybe there's also an opportunity to reuse `formErrorMessage` from `utils/intl`?

<img width="490" alt="image" src="https://github.com/user-attachments/assets/8bf445a9-9bf1-4152-860f-0f21e9f65de5">

"Format JSON" should read "Unformat JSON" when the ABI has been formatted:
<img width="497" alt="image" src="https://github.com/user-attachments/assets/5bbc0880-0e3f-4de8-a811-88a5e77a4ec8">

I agree with Sam that we need a loading state for the ABI field. I checked with Arren and we can just reuse the existing overlay and show a spinner until we get proper designs for it.

A cancel confirmation is also a good idea, we could reuse the one from action form: 
<img width="500" alt="image" src="https://github.com/user-attachments/assets/637e96f7-f538-4c76-a69b-eb637296095e">

I did not get any errors when there was a network error calling the Arbiscan API:
<img width="1682" alt="image" src="https://github.com/user-attachments/assets/eade1de7-64af-4b8c-9380-7dbc4872960f">


JSON/ABI validation ✅ 
<img width="499" alt="image" src="https://github.com/user-attachments/assets/5a759046-10fe-430c-9aa1-7929a999c32c">
<img width="492" alt="image" src="https://github.com/user-attachments/assets/f67e8488-9895-407f-8276-2bf046f73fff">

Modifying the ABI manually to add/delete a function seems to work well to, apart from the issue Sam mentioned:
<img width="499" alt="image" src="https://github.com/user-attachments/assets/14ef9d64-36be-4bb9-99e3-1b163f48adf8">

Param types validation ✅ 
<img width="489" alt="image" src="https://github.com/user-attachments/assets/25870798-7ec7-4ec6-828f-65ff7725c9d3">
<img width="494" alt="image" src="https://github.com/user-attachments/assets/a9b152a1-f456-460b-b302-aaadc5b9e035">
<img width="495" alt="image" src="https://github.com/user-attachments/assets/50ba26f1-a346-4de3-addd-a242fcf44a87">
<img width="495" alt="image" src="https://github.com/user-attachments/assets/75892d61-3313-435d-9f81-a5513bc2985f">
<img width="497" alt="image" src="https://github.com/user-attachments/assets/7537f67e-252c-4729-83de-4f690ca764c4">
<img width="496" alt="image" src="https://github.com/user-attachments/assets/b467c862-3113-4167-8dc3-a45ce4711100">
<img width="488" alt="image" src="https://github.com/user-attachments/assets/9a35a2c2-4bb8-4e21-ae4d-87916df84781">
<img width="490" alt="image" src="https://github.com/user-attachments/assets/9b11b2de-a723-4bb9-97fb-1375fbc2767d">
<img width="490" alt="image" src="https://github.com/user-attachments/assets/aa81e3a4-92a9-4c8b-a432-0eafbb1b7b17">
<img width="492" alt="image" src="https://github.com/user-attachments/assets/9180666a-fb30-4e10-9639-573f8869dccc">
<img width="494" alt="image" src="https://github.com/user-attachments/assets/477a67da-3da3-4b6f-a672-1ce476affafd">
<img width="492" alt="image" src="https://github.com/user-attachments/assets/9eb60740-d809-4b85-87f6-2da3d4fbcc60">
<img width="483" alt="image" src="https://github.com/user-attachments/assets/1436cbfe-e092-418f-bc4a-1b82583b0b1e">
<img width="488" alt="image" src="https://github.com/user-attachments/assets/efedb34a-8c6e-4b6b-b967-da194faf217f">
<img width="493" alt="image" src="https://github.com/user-attachments/assets/151ed77a-1357-42f4-ba3d-e42e0d0dfca3">
<img width="494" alt="image" src="https://github.com/user-attachments/assets/94c50e80-005c-4e53-a581-8523eca98586">
<img width="487" alt="image" src="https://github.com/user-attachments/assets/200ebc07-41f6-441b-adf6-e487c4f6e3ff">
<img width="492" alt="image" src="https://github.com/user-attachments/assets/24c072e3-a166-41da-b0b6-5c7af3efd4cc">
<img width="497" alt="image" src="https://github.com/user-attachments/assets/92af34ce-7aa9-4481-a9ab-83ab5634d21d">
<img width="484" alt="image" src="https://github.com/user-attachments/assets/c1fd46d8-f765-49dd-8c51-2adf54afb3e9">
<img width="491" alt="image" src="https://github.com/user-attachments/assets/f2023f03-89ab-4ed4-9ca7-272cb5c2fb3c">
<img width="492" alt="image" src="https://github.com/user-attachments/assets/b0158a4b-01ff-4db8-8d1d-7329d96c925d">

Nortsova commented on src/components/v5/common/ActionSidebar/partials/forms/ArbitraryTxsForm/partials/AddTransactionModal/consts.ts:
```diff
@@ -0,0 +1,114 @@
+
+import { MSG } from './translation.ts';
+
+const createValidator = (schema: any) => (value: string) => {
+  try {
+    schema.validateSync(value);
+    return true;
+  } catch (error) {
+    return error.message;
+  }
+};
```
I’m not sure if I can, as we’re dealing with a dynamic form here. I’ll investigate what we can do about it. However, from what I see now, we’d need to update the schema every time we regenerate the dynamic fields. This approach seems more complex than simply implementing inline validation, even though it’s inconsistent with the other forms.

We’re still using Yup here, though.

Nortsova commented on src/components/v5/common/ActionSidebar/partials/forms/ArbitraryTxsForm/consts.ts:
```diff
@@ -15,8 +15,6 @@ export const validationSchema = object()
             contractAddress: string().defined(),
```
I am not sure because it is two separate forms; one is the Motion form, another is the Modal form.

Nortsova commented:
Thank you, @iamsamgibbs and @jakubcolony, for your review! I really appreciate it 😃 

Changes since last time:
- `fetchArbiscanContract` deleted and replaced with `fetchContractABI`
- Error formatting and server error handling fixed:
<img width="476" alt="image" src="https://github.com/user-attachments/assets/b3470482-2455-4be4-9d43-e700fbe75b05">

- Confirmation modal added:
<img width="561" alt="image" src="https://github.com/user-attachments/assets/35828d15-09fc-465b-9cf1-27649e98c204">

- Modal buttons fixed for mobile view (Previously, buttons were not sticking to the bottom)
<img width="617" alt="image" src="https://github.com/user-attachments/assets/f1935d65-d739-4575-99a1-043ca315486b">

- Default state for the table set to empty table (This was based on Kerry’s comment https://github.com/JoinColony/colonyCDapp/pull/3624#pullrequestreview-2435351878 and a discussion with @melyndav )
<img width="699" alt="image" src="https://github.com/user-attachments/assets/5ed32e07-7d6a-4fe2-9179-e08abf242f7f">

-Dynamic validation bug fixed (as noted by @iamsamgibbs ):
![Nov-18-2024 21-51-52](https://github.com/user-attachments/assets/9a64582c-f6ea-4056-8571-30b4559c1a18)

- Form fields now clear correctly when the method is changed:
<img width="700" alt="image" src="https://github.com/user-attachments/assets/d369f037-7c76-49c9-8510-c7230750dd4e">

-Spinner added for contract loading:
<img width="465" alt="image" src="https://github.com/user-attachments/assets/c30f23ea-2cf3-4d91-b9f5-9447ca3f1c51">

- Format JSON / Stringify JSON text was added: 
<img width="463" alt="image" src="https://github.com/user-attachments/assets/15083bde-169f-416e-aa82-d9f9e56aeab3">

- Improved filtering. Now only functions (not constants) are displayed:
<img width="517" alt="image" src="https://github.com/user-attachments/assets/dc754a1b-fcc2-44e7-9199-f74589d1f750">

### Remaining Work:
The only open point is validationSchema vs. inline validation. As I mentioned in my earlier comment, inline validation is much more suitable than using a general schema in the case of dynamic fields. However, I will investigate further and get additional input from the team in our chat 🙌 

Thank you ✨

Nortsova commented:
My latest changes are related to this PR: https://github.com/JoinColony/colonyCDapp/pull/3736 . 

I switched from this approach:
```js
const contractAddressField = watch('contractAddress');
useEffect(() => {
  // do something when contractAddressField changes
}, [contractAddressField]);
```

To this: 

```js
  useEffect(() => {
    const { unsubscribe } = watch(
      ({ contractAddress }, { name }) => {
        if (name === 'contractAddress' ) {
          // do something when contractAddress is updated
        }
      },
    );
    return () => unsubscribe();
   }, [watch]);
```
<img width="854" alt="image" src="https://github.com/user-attachments/assets/86527388-a2ba-40cc-bad8-15d5f5b74aab">

The reason for this change is to ensure the effect triggers only when the value is updated and not during the initial rendering. This will be particularly useful when I add "edit" and "defaultValues" functionality in https://github.com/JoinColony/colonyCDapp/pull/3736 . 🙌

iamsamgibbs reviewed (changes requested):
Nice job! I only have one issue, otherwise everything is great.

If I submit an empty form, the table is missing an error validation (and transactions should have a capital T)
 
<img width="678" alt="Screenshot 2024-11-21 at 12 02 04" src="https://github.com/user-attachments/assets/cc25879f-bafc-4eae-80a8-556c6ac3fb7a">

Compared to advanced payments where the table border and contents goes red:

<img width="677" alt="Screenshot 2024-11-21 at 12 02 25" src="https://github.com/user-attachments/assets/2191d7c1-f6e4-4521-ad7f-2b1defadf92a">

Everything else is good though! Better messaging here:

<img width="509" alt="Screenshot 2024-11-21 at 12 02 50" src="https://github.com/user-attachments/assets/7bbd8bcd-2b0c-43d1-8b48-f178f5eb943b">

Validation is updating correctly here:

<img width="513" alt="Screenshot 2024-11-21 at 12 07 00" src="https://github.com/user-attachments/assets/0ae84dba-3cd2-4b5a-94f6-a7624f1eaab6">

Loading spinners look great.

<img width="435" alt="Screenshot 2024-11-21 at 12 10 14" src="https://github.com/user-attachments/assets/181cda22-4e4d-44a9-abcd-45f2bd4cb78f">

Smashed it!

mmioana reviewed (changes requested):
Amazing job on this complex issue @Nortsova 💯 

Went through the testing steps you provided and it seems to be (almost) working as expected

`Modal matches the design`
![Screenshot 2024-11-21 at 14 39 05](https://github.com/user-attachments/assets/b9dc57d9-3fd7-404d-871f-3b8422d6e509)

`Fields are required`
![Screenshot 2024-11-21 at 14 39 16](https://github.com/user-attachments/assets/bd806b37-7ba8-4d3a-822b-ecc5de69378c)

`Invalid contract address`
![Screenshot 2024-11-21 at 14 39 30](https://github.com/user-attachments/assets/4e230b09-7b65-47d8-9e25-7732d8ffe224)

`Unverified contract address`
![Screenshot 2024-11-21 at 14 39 45](https://github.com/user-attachments/assets/9450e809-797b-41ed-b6a1-b0501f30b969)

`Empty contract address`
![Screenshot 2024-11-21 at 14 40 03](https://github.com/user-attachments/assets/69390b92-a36f-4c45-b53e-5881a3b6f3bd)

`Format JSON toggling`

https://github.com/user-attachments/assets/bd4db747-79b8-41ce-84ed-25177302e27f

`Edit generated ABI overlay`

https://github.com/user-attachments/assets/b71a3441-f918-4a85-8cbb-a410dfd14ee0

`Correct contract address`
![Screenshot 2024-11-21 at 14 41 24](https://github.com/user-attachments/assets/f8bd3b3a-e3d2-4a3a-9e64-7ae76e575a55)

`Submitting form triggers validation`
![Screenshot 2024-11-21 at 14 43 09](https://github.com/user-attachments/assets/09ab7b29-cea4-4e4d-a631-9b6da590c4db)

`Selected setAuthority and triggered form validation`
![Screenshot 2024-11-21 at 14 43 31](https://github.com/user-attachments/assets/755fbb31-b452-4ea6-b5af-7fc94a00e8c8)

`Tried a few different scenarios for the authority_ input type`
![Screenshot 2024-11-21 at 14 46 22](https://github.com/user-attachments/assets/1d30f430-a423-4c8f-918a-c2226420f862)
![Screenshot 2024-11-21 at 14 46 28](https://github.com/user-attachments/assets/1d0c7570-3381-4fca-b004-f3b916d522fe)
![Screenshot 2024-11-21 at 14 49 19](https://github.com/user-attachments/assets/55e81658-cef0-4754-a43b-38adfc2ee8a1)
![Screenshot 2024-11-21 at 14 49 49](https://github.com/user-attachments/assets/dd43106f-02b3-4d32-be99-5bec4d7b8fbd)


![Screenshot 2024-11-21 at 14 53 00](https://github.com/user-attachments/assets/5c4f966f-c8a3-475e-9d95-f348a2a3a5e6)
![Screenshot 2024-11-21 at 14 53 07](https://github.com/user-attachments/assets/9f63923b-98eb-4382-b0fd-ddcfbcb7acfd)
![Screenshot 2024-11-21 at 14 53 57](https://github.com/user-attachments/assets/0c06194f-2b17-4f06-981c-7afd9a6ca30e)
![Screenshot 2024-11-21 at 14 54 52](https://github.com/user-attachments/assets/d7a6e801-633f-416e-be2a-7926ab311015)

`Confirmed the transaction`
![Screenshot 2024-11-21 at 14 55 32](https://github.com/user-attachments/assets/c5965b91-c2c5-47c6-9305-42d3b9f0d442)

However @Nortsova I noticed a few misalignments with `Figma` and a bug - though let me know if you'd like to treat these in a separate PR 🙌 

1. The `Edit generated ABI` button styling

This PR
![Screenshot 2024-11-21 at 14 44 12](https://github.com/user-attachments/assets/8b63459d-bcf5-4558-85e0-5345d750a9a2)
While `Figma`
![Screenshot 2024-11-21 at 14 45 37](https://github.com/user-attachments/assets/00e69e59-4749-44c9-9dd0-4e973437af31)

2. Error message for other non-contract addresses

This PR
![Screenshot 2024-11-21 at 14 44 56](https://github.com/user-attachments/assets/71b14b04-a3b3-4d87-acb3-68120f9b394b)
While `Figma`
![Screenshot 2024-11-21 at 14 45 11](https://github.com/user-attachments/assets/2d1a0a14-5ff7-4c3f-87c6-4279abb5ae9d)

And another issue I found while testing was by entering `uint12345678` which completely blocked the UI. Maybe we can add some validation for supported `int{m}` and `uint{m}` values
![Screenshot 2024-11-21 at 14 50 01](https://github.com/user-attachments/assets/99c51656-974d-420f-9b64-bd61a50d4016)

https://github.com/user-attachments/assets/569cd539-c0a6-45ac-9b76-42822dbe2b8a

Nonetheless amazing job on this one 🥇

Nortsova commented:
Thank you, guys. I did the fixes:
<img width="547" alt="image" src="https://github.com/user-attachments/assets/b36fc43d-c5ef-4fdb-87a2-cacbb94cab1e">
<img width="249" alt="image" src="https://github.com/user-attachments/assets/c641abc0-802f-4a17-9bf3-184e233f2a2d">
<img width="468" alt="image" src="https://github.com/user-attachments/assets/2ad67ea6-384f-4551-8439-5155cdde19b0">

Regarding the last @mmioana comment about uint/int validation, I made some fixes to avoid app crashes, but I will speak to @jakubcolony because I believe we need to validate edited ABI for correct types.

jakubcolony reviewed (changes requested):
Thank you for all the changes @Nortsova, it looks like we're nearly there 🎉 

A couple of things I spotted:

When there are no transactions, the UI shows both the empty table and Add transaction button:
<img width="699" alt="image" src="https://github.com/user-attachments/assets/3f36f418-5897-4f62-b7ff-bffb31fd75c8">

Whereas in Figma only the Add transaction button is shown:
<img width="1006" alt="image" src="https://github.com/user-attachments/assets/61c0013a-303d-4749-848d-6946875e5680">

----

Cancel confirmation modal appears even if no contract address/custom ABI was entered:
<img width="496" alt="image" src="https://github.com/user-attachments/assets/66f59d1b-1459-4835-9f8c-1a0ff074b8c9">

----

Network errors are now handled nicely, however when I start editing ABI manually I should not see the "Edit generated ABI" overlay:

https://github.com/user-attachments/assets/1ad31a5b-ab84-4ed6-8feb-5c69b69681f7

----

Adding functions without any arguments triggers validation error in the main form:
<img width="702" alt="image" src="https://github.com/user-attachments/assets/43260013-ea62-442c-8af3-8a6def859600">


----

Loading state looks great, just note we might have to change it once the designs are available ✅ 
https://github.com/user-attachments/assets/17de8782-ed50-439d-9b33-97ea6e517bd0

----

Format JSON changes to Unformat JSON ✅ 
<img width="497" alt="image" src="https://github.com/user-attachments/assets/aeff5504-dff6-4360-bba7-20b19b893c43">

Nortsova commented:
Thank you @jakubcolony . The latest fixes:
- Cancel confirmation modal shows only if form is dirty
![dirty-confirmation](https://github.com/user-attachments/assets/86bdeb8f-5b3b-4d7c-9a8b-627613968fa5)

- ABI "Edit warning" now shows only after the ABI field was uploaded
![abi-warning](https://github.com/user-attachments/assets/7e21a733-48b8-4040-b95e-a3a918fc0f4d)

- Validation bug for empty args fixed

- `SUPPORTED_SAFE_NETWORKS` replaced with `ARBITRARY_TRANSACTION_NETWORKS`

Note: Transactions design was changed a little bit and now we have empty table as well 🙌

jakubcolony reviewed (approved):
Thank you for all the changes, this is now working great! 💯 

Can confirm the previous issues have been fixed:

- Cancel confirmation showing on non-dirty modal:
![cancel-confirm-fix](https://github.com/user-attachments/assets/f65cf8f6-8421-4090-945b-723e37c50d9e)

- "Edit generated ABI" showing on user modified ABI:
![no-overlay-fix](https://github.com/user-attachments/assets/1d38b763-8f59-4906-bca2-52abdacaed41)

- Empty array causing validation errors:

![empty-array-fix](https://github.com/user-attachments/assets/10787d66-16e5-4112-b156-7d865a70a17b)

iamsamgibbs reviewed (approved):
Nice job! This is all still looking and working great.

Modal only shows when form is dirty.

https://github.com/user-attachments/assets/0d696c0f-14b9-49a5-904f-559965a42e51

Table validation here is much better.

<img width="670" alt="Screenshot 2024-11-29 at 11 02 25" src="https://github.com/user-attachments/assets/d1731239-55ee-4c96-91b5-a71c3743e0ff">

(Still missing the capital `T` at the start of transactions in the validation message, but a load of the validation messages across all actions are missing capital letters, so let's just fix this later and get it merged now!)

mmioana commented on src/components/v5/common/ActionSidebar/partials/forms/ArbitraryTxsForm/partials/AddTransactionModal/JsonAbiInput.tsx:
```diff
@@ -0,0 +1,88 @@
+        <div className="absolute bottom-0 left-0 right-0 top-0 flex cursor-pointer items-center justify-center rounded bg-base-sprite opacity-0 transition-opacity hover:opacity-100">
+          <Button
+            onClick={toggleShowEditWarningOff}
+            className="min-w-[12.5rem]"
```
@Nortsova as a minor improvement to correctly match `Figma` please add here `hover:border-gray-300` - but this can be done also in a subsequent PR 🫶

mmioana reviewed (approved):
Amazing work on this PR @Nortsova 👏🥇

The error handling is so much smoother 💯 

Creating an `Arbitrary transaction`
![Screenshot 2024-11-30 at 18 56 21](https://github.com/user-attachments/assets/f8f34586-0ab9-48d3-a28a-9fd234200267)

The transaction modal
![Screenshot 2024-11-30 at 18 56 24](https://github.com/user-attachments/assets/b9da3fde-6096-4806-9675-d659fdd2894d)

When submitting it with empty fields, the errors messages show up
![Screenshot 2024-11-30 at 18 56 37](https://github.com/user-attachments/assets/9ebda6b7-ac08-43e5-b2a9-cdc6ae124e89)

Entering an invalid contract address
![Screenshot 2024-11-30 at 18 56 59](https://github.com/user-attachments/assets/50411bbb-6b96-4943-b703-4dd7c6c5f14f)

Entering a not verified contract address
![Screenshot 2024-11-30 at 18 57 14](https://github.com/user-attachments/assets/16e4eaa8-7ce2-4250-8cb0-315cfdf1b061)

Entering a contract with no methods
![Screenshot 2024-11-30 at 18 57 26](https://github.com/user-attachments/assets/03c79636-882a-4de4-bafb-552a22cb32b4)

The layover shows up upon hovering
![Screenshot 2024-11-30 at 19 01 10](https://github.com/user-attachments/assets/6f211427-3d00-465f-a1ac-6a06357393aa)

Now with a valid contract address
![Screenshot 2024-11-30 at 19 01 52](https://github.com/user-attachments/assets/dcf94c27-3400-45d5-967f-c7f770c24a28)

The layover shows up
![Screenshot 2024-11-30 at 19 02 03](https://github.com/user-attachments/assets/05f4d33c-6c20-4be2-9cba-2131f50d04d0)

Trying to submit before selecting a method
![Screenshot 2024-11-30 at 19 02 15](https://github.com/user-attachments/assets/84a0caa2-f35c-40fc-8ec1-3f93b3de4681)

Trying to submit before entering values for the method's inputs
![Screenshot 2024-11-30 at 19 02 38](https://github.com/user-attachments/assets/30adc0fa-7924-478a-a227-f3384d463cc1)

Now tested different input types and values
![Screenshot 2024-11-30 at 19 02 50](https://github.com/user-attachments/assets/a8c92eb5-1f38-41cb-8f26-47201dbb2849)
![Screenshot 2024-11-30 at 19 03 15](https://github.com/user-attachments/assets/945ab516-3351-43bc-94a6-fb9ef8a33844)
![Screenshot 2024-11-30 at 19 03 43](https://github.com/user-attachments/assets/db52cb48-6be4-42be-8da0-5326134c97e7)
![Screenshot 2024-11-30 at 19 03 48](https://github.com/user-attachments/assets/d3071ad5-f014-4ab6-8ad5-5224cc254180)
![Screenshot 2024-11-30 at 19 05 01](https://github.com/user-attachments/assets/8db4a89a-3fe1-4bf9-9a19-c3ba2d8a64ad)
![Screenshot 2024-11-30 at 19 06 52](https://github.com/user-attachments/assets/efe18cf1-88f1-48d8-835f-5ea6919b5b65)
![Screenshot 2024-11-30 at 19 07 37](https://github.com/user-attachments/assets/e21dd822-36ac-43b7-a758-273102f8612a)
![Screenshot 2024-11-30 at 19 07 43](https://github.com/user-attachments/assets/df707a97-4da2-4b20-9013-5a71d774b9a1)

https://github.com/user-attachments/assets/ed96c297-19f7-46fc-9942-6426eac0b674

And managed to submit a transaction
![Screenshot 2024-11-30 at 19 13 06](https://github.com/user-attachments/assets/d48bb0a8-2de9-4fa4-b429-8a38ec4e6aaf)

And submit the motion
![Screenshot 2024-11-30 at 19 13 21](https://github.com/user-attachments/assets/95b845e7-d494-4c3f-a42a-4974098e0938)

Great work once again ❤️
